File: java/lru_cache/src/main/java/com/qe/test/cache/LRUCache.java
package com.qe.test.cache;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.BiConsumer;

public class LRUCache<K, V> {
    private final int capacity;
    private final boolean threadSafe;
    private final Map<K, Node<K, V>> map;
    private final Node<K, V> head;
    private final Node<K, V> tail;
    private final ReentrantLock lock;

    public LRUCache(int capacity) {
        this(capacity, false);
    }

    public LRUCache(int capacity, boolean threadSafe) {
        if (capacity <= 0) throw new IllegalArgumentException("capacity must be positive");
        this.capacity = capacity;
        this.threadSafe = threadSafe;
        this.map = new HashMap<>(capacity);
        this.lock = threadSafe ? new ReentrantLock() : null;
        this.head = new Node<>(null, null);
        this.tail = new Node<>(null, null);
        head.next = tail;
        tail.prev = head;
    }

    public V get(K key) {
        if (threadSafe) {
            lock.lock();
            try {
                return getInternal(key);
            } finally {
                lock.unlock();
            }
        }
        return getInternal(key);
    }

    public void put(K key, V value) {
        Objects.requireNonNull(key, "key must not be null");
        if (threadSafe) {
            lock.lock();
            try {
                putInternal(key, value);
            } finally {
                lock.unlock();
            }
        } else {
            putInternal(key, value);
        }
    }

    public boolean contains(K key) {
        if (threadSafe) {
            lock.lock();
            try {
                return map.containsKey(key);
            } finally {
                lock.unlock();
            }
        }
        return map.containsKey(key);
    }

    public int size() {
        if (threadSafe) {
            lock.lock();
            try {
                return map.size();
            } finally {
                lock.unlock();
            }
        }
        return map.size();
    }

    public void forEach(BiConsumer<? super K, ? super V> action) {
        Objects.requireNonNull(action);
        if (threadSafe) {
            lock.lock();
            try {
                forEachInternal(action);
            } finally {
                lock.unlock();
            }
        } else {
            forEachInternal(action);
        }
    }

    private V getInternal(K key) {
        var node = map.get(key);
        if (node == null) return null;
        moveToHead(node);
        return node.value;
    }

    private void putInternal(K key, V value) {
        var existing = map.get(key);
        if (existing != null) {
            existing.value = value;
            moveToHead(existing);
            return;
        }
        if (map.size() >= capacity) {
            evict();
        }
        var node = new Node<>(key, value);
        map.put(key, node);
        addToHead(node);
    }

    private void evict() {
        var lru = tail.prev;
        if (lru == head) return;
        removeNode(lru);
        map.remove(lru.key);
    }

    private void moveToHead(Node<K, V> node) {
        removeNode(node);
        addToHead(node);
    }

    private void addToHead(Node<K, V> node) {
        node.next = head.next;
        node.prev = head;
        head.next.prev = node;
        head.next = node;
    }

    private void removeNode(Node<K, V> node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private void forEachInternal(BiConsumer<? super K, ? super V> action) {
        var cur = head.next;
        while (cur != tail) {
            action.accept(cur.key, cur.value);
            cur = cur.next;
        }
    }

    private static class Node<K, V> {
        final K key;
        V value;
        Node<K, V> prev;
        Node<K, V> next;

        Node(K key, V value) {
            this.key = key;
            this.value = value;
        }
    }
}

File: python/lru_cache/implementation/lru_cache.py
from __future__ import annotations

import threading
from typing import Generic, TypeVar, cast

KT = TypeVar("KT")
VT = TypeVar("VT")


class _Node(Generic[KT, VT]):
    __slots__ = ("key", "value", "prev", "next")

    def __init__(self, key: KT, value: VT) -> None:
        self.key = key
        self.value = value
        self.prev: _Node[KT, VT] | None = None
        self.next: _Node[KT, VT] | None = None


class LRUCache(Generic[KT, VT]):
    def __init__(self, capacity: int, thread_safe: bool = False) -> None:
        if capacity < 1:
            raise ValueError("capacity must be >= 1")
        self._capacity = capacity
        self._cache: dict[KT, _Node[KT, VT]] = {}
        self._head = _Node(cast(KT, None), cast(VT, None))
        self._tail = _Node(cast(KT, None), cast(VT, None))
        self._head.next = self._tail
        self._tail.prev = self._head
        self._lock = threading.Lock() if thread_safe else None

    def _acquire(self) -> None:
        if self._lock is not None:
            self._lock.acquire()

    def _release(self) -> None:
        if self._lock is not None:
            self._lock.release()

    def _remove_node(self, node: _Node[KT, VT]) -> None:
        prev_node = node.prev
        next_node = node.next
        if prev_node is not None:
            prev_node.next = next_node
        if next_node is not None:
            next_node.prev = prev_node

    def _add_to_front(self, node: _Node[KT, VT]) -> None:
        node.prev = self._head
        node.next = self._head.next
        if self._head.next is not None:
            self._head.next.prev = node
        self._head.next = node

    def get(self, key: KT) -> VT | None:
        self._acquire()
        try:
            node = self._cache.get(key)
            if node is None:
                return None
            self._remove_node(node)
            self._add_to_front(node)
            return node.value
        finally:
            self._release()

    def put(self, key: KT, value: VT) -> None:
        self._acquire()
        try:
            if key in self._cache:
                node = self._cache[key]
                node.value = value
                self._remove_node(node)
                self._add_to_front(node)
                return
            if len(self._cache) >= self._capacity:
                lru = self._tail.prev
                if lru is not None and lru is not self._head:
                    self._remove_node(lru)
                    del self._cache[lru.key]
            new_node = _Node(key, value)
            self._cache[key] = new_node
            self._add_to_front(new_node)
        finally:
            self._release()

    @property
    def capacity(self) -> int:
        return self._capacity

    def __len__(self) -> int:
        self._acquire()
        try:
            return len(self._cache)
        finally:
            self._release()
